crates/datasources/src/sqlserver/client.rs

```diff
@@ -0,0 +1,209 @@
+/// while the Connection is what's actually used to communicate with the server.
+/// The Connection should be handled in a background thread, and should run to
+/// completion.
+pub async fn connect<S>(config: tiberius::Config, stream: S) -> Result<(Client, Connection<S>)>
+where
+    S: AsyncRead + AsyncWrite + Unpin + Send,
+{
+    let (sender, receiver) = mpsc::unbounded_channel();
+    let client = tiberius::Client::connect(config, stream).await?;
+
+    Ok((Client::new(sender), Connection::new(client, receiver)))
+}
+
+/// Wrapper around the tiberius client.
+#[derive(Debug)]
+pub struct Connection<S: AsyncRead + AsyncWrite + Unpin + Send> {
+    /// The actual client to the server.
+    client: tiberius::Client<S>,
+    /// Channel for requests from client.
+    receiver: mpsc::UnboundedReceiver<Request>,
+}
+
+impl<S> Connection<S>
+where
+    S: AsyncRead + AsyncWrite + Unpin + Send,
+{
+    fn new(client: tiberius::Client<S>, receiver: mpsc::UnboundedReceiver<Request>) -> Self {
+        Connection { client, receiver }
+    }
+
+    /// Runs the connection to completion.
+    ///
+    /// Incoming messages from the client will be processed to completion before
+    /// moving on to the next message.
+    pub async fn run(mut self) -> Result<()> {
+        while let Some(req) = self.receiver.recv().await {
+            match req {
+                Request::Query { query, response } => {
+                    let mut stream = match self.client.query(query, &[]).await {
+                        Ok(stream) => stream,
+                        Err(e) => {
+                            // We don't care if this errors, just means that the
+                            // client is no longer listening.
+                            let _ = response.send(Err(e.into())).await;
+                            continue;
+                        }
+                    };
+
+                    while let Some(item) = stream.next().await {
+                        if response.send(item.map_err(|e| e.into())).await.is_err() {
+                            // Client no longer listening, stop reading from the
+                            // stream and just move to the next message.
+                            break;
+                        }
+                    }
+                }
+                Request::Drop => {
+                    debug!("closing SQL Server connection");
+                    self.client.close().await?;
+                    return Ok(());
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Client side of the connection.
+#[derive(Debug)]
+pub struct Client {
+    sender: mpsc::UnboundedSender<Request>,
+}
+
+impl Client {
+    fn new(sender: mpsc::UnboundedSender<Request>) -> Self {
+        Client { sender }
+    }
+
+    // TODO: Zero clue how we'd want to handle params. Might as well just fork
+    // tiberius to make it more reasonable.
```

Comment:
We are not using a connection pool. There's a one to one mapping between `Client` and `Connection` (which maps to a single session on the sql server side). Currently we're only sending a single message at a time. The `Client` _can_ handle multiple messages, but the implementation of `run` means one message will be handled entirely to completion before working on the next one. Also `Client` isn't clone, so it'd be very hard to actually do that.

If/when we do connection caching, I can see this becoming something we'll want to look at further.

Reply:
makes sense; the client connection pooling/caching is probably going to get us if people end up using glaredb as a passthrough for some operational workloads at any point, but for now it seems fine.

Reply:
My idea for connection caching would be to do caching per session, and since session only handle one query at a time, I don't see this being too big of an issue. There might be issues if querying multiple tables from the same sql server (as querying one would block querying the other) but I think that's something we can figure out when we get there.